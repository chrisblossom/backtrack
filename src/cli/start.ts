import { optionsFile } from '../options-file/options-file';
import { loadStatsFile } from '../stats-file/load-stats-file';
import { writeStatsFile } from '../stats-file/write-stats-file';
import { handleError } from '../utils/handle-error';
import { run } from '../run/run';
import { clean } from '../clean/clean';
import { fileManager } from '../file-manager/file-manager';
import { packageJsonManager } from '../package-json-manager/package-json-manager';
import log from '../utils/log';

async function start(options?: {}) {
	const taskName =
		process.env.RUN_MODE && typeof process.env.RUN_MODE === 'string'
			? process.env.RUN_MODE
			: 'cli-start';

	if (!process.env.NODE_ENV) {
		process.env.NODE_ENV = 'development';
	}

	const NODE_ENV = process.env.NODE_ENV;
	const logPrefix = `${taskName}:${NODE_ENV}`;

	try {
		const lifecycles = optionsFile();
		const statsFile = loadStatsFile();

		/**
		 * Update package.json
		 */
		const packageJsonStats = await packageJsonManager(
			lifecycles,
			statsFile.packageJson,
		);

		/**
		 * Update all files
		 */
		const filesStats = await fileManager(
			lifecycles.files,
			statsFile.fileManager,
		);

		/**
		 * Write stats file
		 */
		await writeStatsFile(
			{
				fileManager: filesStats,
				packageJson: packageJsonStats,
			},
			statsFile,
		);

		if (
			process.env.RUN_MODE === '--init' ||
			process.env.RUN_MODE === 'init'
		) {
			/**
			 * Run clean script initially
			 */
			if (lifecycles.clean) {
				await clean(lifecycles.clean);
			}

			log.success('backtrack initialized');

			return;
		}

		const matchedTask = lifecycles[taskName];
		if (!matchedTask || matchedTask.length === 0) {
			throw new Error(`${taskName} not found in preset`);
		}

		await run(taskName, matchedTask, options);
	} catch (error) {
		await handleError({
			error,
			logPrefix,
		});
	}
}

export { start };
