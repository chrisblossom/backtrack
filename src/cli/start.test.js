/* @flow */

import path from 'path';

require('./start');

const start = () => require('./start').start();

describe('start', () => {
    const cwd = process.cwd();
    let RUN_MODE;
    let run;

    beforeEach(() => {
        jest.mock('../utils/log', () => ({
            warn: jest.fn(),
            info: jest.fn(),
            error: jest.fn(),
            success: jest.fn(),
        }));

        // $FlowIssue
        run = require.requireMock('../run/run').run;
        jest.mock('../run/run.js', () => ({ run: jest.fn() }));
        jest.mock('../utils/handle-error.js', () => ({
            handleError: jest.fn(),
        }));

        RUN_MODE = process.env.RUN_MODE;
    });

    afterEach(() => {
        process.chdir(cwd);

        process.env.RUN_MODE = RUN_MODE;
    });

    test('runs task', async () => {
        process.env.RUN_MODE = 'dev';

        const dir = path.resolve(__dirname, '__sandbox__/app1/');
        process.chdir(dir);

        await start();

        expect(run.mock.calls).toMatchSnapshot();
    });
});
