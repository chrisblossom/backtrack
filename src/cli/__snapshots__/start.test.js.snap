// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`start runs task 1`] = `
Array [
  Array [
    "dev",
    Array [
      [Function anonymous],
      [Function anonymous],
    ],
    undefined,
  ],
]
`;
